File: src/backend/mdfe/src/model/mdfe_emitente_model.py
from src import db


class MdfeEmitenteModel(db.Model):
    __tablename__ = 'mdfe_emitente'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(60))
    fantasia = db.Column(db.String(60))
    cnpj = db.Column(db.String(14))
    ie = db.Column(db.Integer)
    logradouro = db.Column(db.String(60))
    numero = db.Column(db.String(60))
    complemento = db.Column(db.String(60))
    bairro = db.Column(db.String(60))
    codigo_municipio = db.Column(db.String(7))
    nome_municipio = db.Column(db.String(60))
    cep = db.Column(db.String(8))
    uf = db.Column(db.String(2))
    telefone = db.Column(db.String(12))
    email = db.Column(db.String(60))
    id_mdfe_cabecalho = db.Column(db.Integer, db.ForeignKey('mdfe_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_mdfe_cabecalho = data.get('idMdfeCabecalho')
        self.nome = data.get('nome')
        self.fantasia = data.get('fantasia')
        self.cnpj = data.get('cnpj')
        self.ie = data.get('ie')
        self.logradouro = data.get('logradouro')
        self.numero = data.get('numero')
        self.complemento = data.get('complemento')
        self.bairro = data.get('bairro')
        self.codigo_municipio = data.get('codigoMunicipio')
        self.nome_municipio = data.get('nomeMunicipio')
        self.cep = data.get('cep')
        self.uf = data.get('uf')
        self.telefone = data.get('telefone')
        self.email = data.get('email')

    def serialize(self):
        return {
            'id': self.id,
            'idMdfeCabecalho': self.id_mdfe_cabecalho,
            'nome': self.nome,
            'fantasia': self.fantasia,
            'cnpj': self.cnpj,
            'ie': self.ie,
            'logradouro': self.logradouro,
            'numero': self.numero,
            'complemento': self.complemento,
            'bairro': self.bairro,
            'codigoMunicipio': self.codigo_municipio,
            'nomeMunicipio': self.nome_municipio,
            'cep': self.cep,
            'uf': self.uf,
            'telefone': self.telefone,
            'email': self.email,
        }

File: src/frontend/mdfe/lib/app/data/model/mdfe_emitente_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:mdfe/app/infra/infra_imports.dart';
import 'package:mdfe/app/data/domain/domain_imports.dart';

class MdfeEmitenteModel {
	int? id;
	int? idMdfeCabecalho;
	String? nome;
	String? fantasia;
	String? cnpj;
	int? ie;
	String? logradouro;
	String? numero;
	String? complemento;
	String? bairro;
	String? codigoMunicipio;
	String? nomeMunicipio;
	String? cep;
	String? uf;
	String? telefone;
	String? email;

	MdfeEmitenteModel({
		this.id,
		this.idMdfeCabecalho,
		this.nome,
		this.fantasia,
		this.cnpj,
		this.ie,
		this.logradouro,
		this.numero,
		this.complemento,
		this.bairro,
		this.codigoMunicipio,
		this.nomeMunicipio,
		this.cep,
		this.uf,
		this.telefone,
		this.email,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'fantasia',
		'cnpj',
		'ie',
		'logradouro',
		'numero',
		'complemento',
		'bairro',
		'codigo_municipio',
		'nome_municipio',
		'cep',
		'uf',
		'telefone',
		'email',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Fantasia',
		'Cnpj',
		'Ie',
		'Logradouro',
		'Numero',
		'Complemento',
		'Bairro',
		'Codigo Municipio',
		'Nome Municipio',
		'Cep',
		'Uf',
		'Telefone',
		'Email',
	];

	MdfeEmitenteModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idMdfeCabecalho = jsonData['idMdfeCabecalho'];
		nome = jsonData['nome'];
		fantasia = jsonData['fantasia'];
		cnpj = jsonData['cnpj'];
		ie = jsonData['ie'];
		logradouro = jsonData['logradouro'];
		numero = jsonData['numero'];
		complemento = jsonData['complemento'];
		bairro = jsonData['bairro'];
		codigoMunicipio = jsonData['codigoMunicipio'];
		nomeMunicipio = jsonData['nomeMunicipio'];
		cep = jsonData['cep'];
		uf = MdfeEmitenteDomain.getUf(jsonData['uf']);
		telefone = jsonData['telefone'];
		email = jsonData['email'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idMdfeCabecalho'] = idMdfeCabecalho != 0 ? idMdfeCabecalho : null;
		jsonData['nome'] = nome;
		jsonData['fantasia'] = fantasia;
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['ie'] = ie;
		jsonData['logradouro'] = logradouro;
		jsonData['numero'] = numero;
		jsonData['complemento'] = complemento;
		jsonData['bairro'] = bairro;
		jsonData['codigoMunicipio'] = codigoMunicipio;
		jsonData['nomeMunicipio'] = nomeMunicipio;
		jsonData['cep'] = Util.removeMask(cep);
		jsonData['uf'] = MdfeEmitenteDomain.setUf(uf);
		jsonData['telefone'] = Util.removeMask(telefone);
		jsonData['email'] = email;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idMdfeCabecalho = plutoRow.cells['idMdfeCabecalho']?.value;
		nome = plutoRow.cells['nome']?.value;
		fantasia = plutoRow.cells['fantasia']?.value;
		cnpj = plutoRow.cells['cnpj']?.value;
		ie = plutoRow.cells['ie']?.value;
		logradouro = plutoRow.cells['logradouro']?.value;
		numero = plutoRow.cells['numero']?.value;
		complemento = plutoRow.cells['complemento']?.value;
		bairro = plutoRow.cells['bairro']?.value;
		codigoMunicipio = plutoRow.cells['codigoMunicipio']?.value;
		nomeMunicipio = plutoRow.cells['nomeMunicipio']?.value;
		cep = plutoRow.cells['cep']?.value;
		uf = plutoRow.cells['uf']?.value != '' ? plutoRow.cells['uf']?.value : 'AC';
		telefone = plutoRow.cells['telefone']?.value;
		email = plutoRow.cells['email']?.value;
	}	

	MdfeEmitenteModel clone() {
		return MdfeEmitenteModel(
			id: id,
			idMdfeCabecalho: idMdfeCabecalho,
			nome: nome,
			fantasia: fantasia,
			cnpj: cnpj,
			ie: ie,
			logradouro: logradouro,
			numero: numero,
			complemento: complemento,
			bairro: bairro,
			codigoMunicipio: codigoMunicipio,
			nomeMunicipio: nomeMunicipio,
			cep: cep,
			uf: uf,
			telefone: telefone,
			email: email,
		);			
	}

	
}